### ISSUE TITLE
Generating needed @localid values after editing

### ISSUE THREAD

**Issue by briesenberg07** [uwlib-cams/map_storage#8]
Hi @mcm104 
*Questions and comments about generating needed local ids after the storage source is edited/properties are "filled in" by editors*

# [`add_localid.py`](https://github.com/uwlib-cams/map_storage/blob/1991fcdd8212f306f47bb40aed935aa1a5eded40/scripts/add_localid.py)
## Questions/comments
- Will the [method for setting the `id_num` var starting value](https://github.com/uwlib-cams/map_storage/blob/1991fcdd8212f306f47bb40aed935aa1a5eded40/scripts/add_localid.py#L18) need to change? 
   - For example, when running the script on a property with an existing `implementationSet` element (with localid) *and* a new `implementationSet` element (without any local id yet)--a storage instance that looks like example [1] below.
- `add_localid.py` doesn't seem to output to `localid_log.xml`--is it still needed? I think the purpose of the log was to ensure that once set, local ids in the source instance remain the same. But perhaps we can be assured that that will be the case using `add_localid.py`? Perhaps related to the question above.
- Eventually, I believe we will want the script for adding local ids to read in a file with name X and output a file with the same name X (not create a new, differently-named file). But lots is still in flux right now*, so this doesn't need to be fixed right now.

## Lower priority
- [L20](https://github.com/uwlib-cams/map_storage/blob/1991fcdd8212f306f47bb40aed935aa1a5eded40/scripts/add_localid.py#L20) generates the localid attribute values with an f-string, I understand everything but the `:03` notation. Is this slice notation?
- That's annoying that the namespaces in the source get overwritten with `ns0`, etc. Thanks for fixing that up.

## Example for [1]
```
        <prop localid_prop="rda_Manifestation_m/P30217">
            <prop_iri iri="http://rdaregistry.info/Elements/m/P30217" />
            <prop_label xml:lang="en">has details of base material</prop_label>
            <prop_domain iri="http://rdaregistry.info/Elements/c/C10007" />
            <platformSet>
                <sinopia>
                    <implementationSet localid_implementationSet="is_P30217_001">
                        <resource mapid_resource="Manifestation">
                            <format mapid_format="monograph" />
                            [...]
                    </implementationSet>
                    <implementationSet localid_implementationSet="">
                        <resource mapid_resource="Manifestation">
                            <format mapid_format="serials"/>
                            [...]
                    </implementationSet>
                </sinopia>
            </platformSet>
        </prop>
```

**Comment by mcm104** [comment]
# Re: id_num method:
The way the code is currently written, when processing the code from your example, for the first set, it would overwrite "is_P30217_001" with "is_P30217_001" -- i.e. it would stay the same. Then for the second set, it would put in "is_P30217_002".

As long as the order of the implementation sets doesn't change -- e.g. if we added a new implementation set above 001 instead of below it -- it should work fine!

If the order of implementation sets may change over time though, then yes, the IDs would have to be saved elsewhere and referenced so they wouldn't change every time. However, since the script is in Python now, this wouldn't be an issue the way it was in XSLT!

But then again, if we want to do this in XSLT eventually once the XML structure is finished-ish, then it could be an issue again.

Let me know if any of that doesn't make sense or if you still have concerns!

# Re: localid_log.xml
You're right that `add_localid.py` does **not** output to `localid_log.xml` anymore -- this is based on the original plan we discussed when I was still planning to write this script in XSLT. In XSLT, keeping a log of local IDs would require an additional step of processing -- once to read the file, once to write to it. `add_localid.py` does not keep a log because the plan is that local IDs will be predictable and therefore not need to be recorded.

However, this is no longer a concern if `add_localid.py` is in Python, as we can read and write to the same file in the same script. So if we want to keep a log of local IDs, we can! But shouldn't have to, unless the order of implementation sets may be unreliable. Either way, we can remove this file for now. If we want to keep a log later, it will have to be regenerated anyway, so there's no need to save this one.

# Re: output of `add_localid.py`
That makes sense! I gave the output file a different name just to make it clear which file had been run through the script and also to prevent overwriting the original file if there were any problems with it, but going forward, I can just output it to the same file.

# Re: :03
This is actually something I'm using for the first time, found from [this Stack Overflow thread](https://stackoverflow.com/questions/339007/how-to-pad-zeroes-to-a-string)!

**Comment by briesenberg07** [comment]
Thank you for the response @mcm104 ! 

Going to check a few things off as I get to them:

**Cleanup related to this issue**
- Delete `localid_log.xml` [in this commit](https://github.com/uwlib-cams/map_storage/commit/88b44d620ba11f0c0bb28279be38cc7d9165afea)
- Delete `record_localid.xsl` [in this commit](https://github.com/uwlib-cams/map_storage/commit/d4277b74187aa5fc342362b465df0224e2b268a4)

**`add_localid.py`**
> we will want the script for adding local ids to read in a file with name X and output a file with the same name X (not create a new, differently-named file)
- Taking a swing at this... NOPE my file-handling skills in Python were not up to the task apparently :(
- So this is still a to-do for me or you (going to read up on file-handling)
- One other note regarding `add_localid.py` is that it seems comments `<!-- like this -->` are eliminated in the output file; we will want to retain the comments we make in the storage source

**Comment by briesenberg07** [comment]
File-naming issue fixed! (Thanks again Melissa.)
Closing this issue, which was a bit too broad anyway (I'm bad about combining lots of issues into one issue).
